### PR TITLE
Build cleanup and CentOS docker rewrite

### PR DESCRIPTION
**Pull request by mdymczyk**
Fixes: #355, #453, #456, #501, #564

This PR does quite a bit of stuff (and looks bigger than it it - most files were just moved arround so don't be intimidated by the number of files changed) that I found hard to do in several PRs (or that would require a lot of useless work if not done together as were tightly couples):

1) rewrite all our docker files as CentOS - this allows us to generate code which will run on most Linux distros (glibc 2.17+) and have only one set of dockers
1a) the CentOS docker was based on an existing CentOS docker but slightly optimized (instead of building by hand some libs they are already in package repos)
2) moved all the build (Makefile) config to `make/config.mk` file for clarity and ease of use
3) remove redundant Make targets - cuda8/9 separate targets are not necessary as it's all decided based on CUDA_HOME anyway, docker targets can be controlled with an env variable instead of repeating same targets etc.
4) remove nonccl builds (but leave the option to build without nccl support if need be) as they are not used 
5) clean stage names
6) move all the Jenkins files into `ci` folder
7) move all the tests into `test` folder and appropriate subfolders
8) move requirements files into `interface_py` subfolder as they are Python related
9) benchmark and AWS Jenkinsfiles temporarily removed (still in git history) and we will need to bring them back asap (especially benchmarks) but neither worked and I don't want to fix them in this PR (nor I want to leave dead code in the repository as people might think it's there because it is working).
10) only 4 builds will remain for now `h2o4gpu-ppc64le-cuda8, h2o4gpu-ppc64le-cuda9, h2o4gpu-x86_64-cuda8, h2o4gpu-x86_64-cuda9` - the rest will be removed after this gets merged in.
11) I'm removing `whl` uploaded files marked with `always upload for testing` in the comments not to pollute our s3
12) using `$(PYTHON)` (and `$(PYTHON) -m pip` in our scripts

Comments welcome.

One more potential change to come: after we upload test data to the ppc64le machine with the GPU it might turn out we can only do `testfast` on it as it's a K80, will see how long the full suite takes.

*EDIT*: temporarily blocked because we lost a ppc64 machine with GPUs for tests.

*EDIT2*: before merging need to double check that:
* all the settings set in `make/config.mk` are propagated to Makefiles
* whls work with other projects

### CODE REVIEW

**Comment by navdeep-G** [comment]
@mdymczyk This looks like a huge improvement!

I can run through some test cases. What are we interested in? One is IBM power. Anything else?

File: Dockerfile-build-centos7.in [hunk 1]
```diff
@@ -1,138 +0,0 @@
-FROM FROM_SUBST
```
Comment (navdeep-G) [inline Dockerfile-build-centos7.in]: One Docker file for all. Love it!

Comment (navdeep-G) [inline Dockerfile-build-centos7.in]: One Docker file for all. Love it!

File: Makefile [hunk 2]
```diff
@@ -1,85 +1,16 @@
...
-S3_CMD_LINE := aws s3
-
-#
-# BUILD_INFO setup
-#
-H2O4GPU_COMMIT ?= $(shell git rev-parse HEAD)
-H2O4GPU_BUILD_DATE := $(shell date)
-H2O4GPU_BUILD ?= "LOCAL BUILD @ $(shell git rev-parse --short HEAD) build at $(H2O4GPU_BUILD_DATE)"
-H2O4GPU_SUFFIX ?= "+local_$(shell git describe --always --dirty)"
+include make/Makefile_header.mk
```
Comment (navdeep-G) [inline Makefile]: Great improvements in terms of removing dead code and simplifying things. I think another good exercise is to review the Makefile and see if we totally understand each target. How much are repeating? Can anything be simplified. I think you have taken a huge step in this approach. Just something for others to keep in mind going forward.

**Comment by navdeep-G** [comment]
I have added some tiny comments. Will try a thorough review today but I think @RAMitchell idea of testing it out is great.

**Comment by mdymczyk** [comment]
@RAMitchell thanks for the TODO catch!

@navdeep-G yes, the Makefiles (all of them) might still require some small (or not so small) tune-up I think we can do them when we finally get to adding Windows support. Also maybe a small reshuffle of the tests would be nice so we have clear "alltests", "smoketests", "fasttests", "slowtests" targets would be nice (now we have a `testfast` target but it's just running hand picked subfolders of `open_data` tests).

**Comment by pseudotensor** [comment]
@mdymczyk If you really did move around files without using git mv, can you fix that up so you use git mv first before editing them?

**Comment by mdymczyk** [comment]
@pseudotensor good point and that's what I've been also worried about but after checking the CentOS repositories it looks like we only have to build 2 things currently:

- git, because the latest in their repo is 1.8.3 and we need 1.8.4 for `submodule update --depth`
- swig, they are using some archaic version

Everything else is pulled through `yum install` .

Libraries for ppc64le are a different story, we need to build pillow and arrow but from what Tom said you can't even install Ubuntu on the new Power9 boxes.

Think the current docker file is pretty similar to the original and that origianally it took Tom so long because he was doing it part time, wasn't really up to speed with the project and the ppc platform was new to us all.

So what I'd suggest is to try these Centos dockers (since we already maintain ubuntu and centos docker anyway) and if they turn out to be a major maintenance issue we can discuss with Tom and maybe revert to Ubuntu.

WDYT?

**Pull request merged**